publicodes/source/index.ts

```diff
@@ -156,4 +162,14 @@ export default class Engine<Names extends string> {
 	getParsedRules(): ParsedRules<Names> {
 		return this.parsedRules
 	}
+
+	evaluateNode(node) {
+		if (!node.nodeKind) {
+			throw Error('The provided node must have a "nodeKind" attribute')
+		} else if (!evaluationFunctions[node.nodeKind]) {
+			throw Error(`Unknown "nodeKind": ${node.nodeKind}`)
+		}
+
+		return evaluationFunctions[node.nodeKind].call(this, node)
```

Comment:
Est-ce considéré comme good practice de faire des `call(this, node)` et de définir des fonctions libres (hors objet/classe) faisant utilisation du `this`?

De mon point de vue c'est net négatif à cause de la lisibilité qui en prend un coup, sans parler du problème de dépendance/contrat du point de vue de la définition de la fonction. En tant que développeur, pour comprendre quel était le `this` que j'ai pu lire dans publicode/source/mecanisms/synchronisation.tsx:evaluate, j'ai dû remonter à la définition de `evaluationFunctions` dans evaluate.ts et ensuite remonter ici dans index.ts

A mon avis il faut être plus patient et verbose dans le nommage de nos variable. Une API pour les fonctions d'évaluation qui serait vraiment plus clean pour moi sans plus d'effort: `myEvalFunc(engine: Engine, node)` avec une claire définition de `Engine` grâce à la classe.

Je suis prêt à proposer un petit refacto rapide pour ça si ça te semble bien @mquandalle

Reply:
Avant cette PR on avait plein de paramètres en argument `(cache, situation, parsedRules, node)` avec le `node` qui nous intéresse vraiment tout au bout, le but était donc de regrouper `cache`, `situation` et `parsedRules` dans un objet context pour avoir une signature `(engine, node)` comme tu le proposes. J'ai ensuite proposé de mettre le engine dans le  `this` pour deux raisons :

* l'API de l'engine est la même que l’interpréteur publicode dans `index.ts`
* le engine doit toujours être passé tel quel aux fonctions enfants appelées par le `evaluate`, ce qui est exactement le comportement du `this`

Peut-être que la signature n'est pas assez explicite ? Parce qu'avec `myEvalFunc(this: Engine, node)` on est vraiment pas loin de ta proposition, et le comportement du this est désiré pour passer le contexte dans tout l'arbre d'évaluation.

Reply:
> Avant cette PR on avait plein de paramètres en argument (cache, situation, parsedRules, node) avec le node qui nous intéresse vraiment tout au bout, le but était donc de regrouper cache, situation et parsedRules dans un objet context pour avoir une signature (engine, node) comme tu le proposes.

Ok pour ça. Réduire les arguments à seulement engine et node va rejeter la responsabilité de gérer le contrat entre `evaluateNode` et les différents `myEvaluate` non plus au niveau des arguments (la signature de la fonction était le contrat) mais au niveau de l'API publique de la classe Engine (à Engine de faire gaffe à  proposer les bonnes méthodes publiques). Je n'ai pas d'avis sur la question, time will tell.

> l'API de l'engine est la même que l’interpréteur publicode dans index.ts

Qu'est-ce que "l'intepréteur publicode" si ce n'est l'instance de l'engine? J'ai l'impression que tu fais une distinction sémantique que je ne fais pas. Peux-tu stp m'expliquer?  
Il semblerait que tu trouves que le `this` dans les fonctions est naturel quand tu fais un interpréteur, mais je te confirmes que ça ne l'est pas du tout pour moi =)

> engine doit toujours être passé tel quel aux fonctions enfants appelées par le evaluate

Je ne suis pas sûr de comprendre le "tel quel": veux-tu dire qu'il ne doit pas s'agit d'un autre engine? En gros que engine est (très grossomodo) un singleton?

> Peut-être que la signature n'est pas assez explicite ? Parce qu'avec myEvalFunc(this: Engine, node) on est vraiment pas loin de ta proposition, et le comportement du this est désiré pour passer le contexte dans tout l'arbre d'évaluation.

Oui en effet ce n'était pas explicite (du tout IMHO, cf ci-dessus). Ok pour `myEvalFunc(this: Engine, node)` même si je trouve que c'est moins explicite que `engine: Engine`.

Si je regarde bêtement le code comme un programmeur qui ne connaît pas du tout publicodes, le seul et unique lien entre ton `this` et l'engine est le `.call(this`. Le `this` est un mal nécessaire dans le cadre du `.call(this`, pour ça ok, on ne peut pas y réchapper. Mais dans le reste du code, c'est une obfuscation de mon point de vue.

Politiquement, je vote à fond pour la lisibilité. Je pense qu'il faut faire plus de sacrifices pour la lisibilité bête et méchante, et la destruction systématique de tout implicite. Le fait qu'on se retrouve à en parler en témoigne :)

Reply:
> Qu'est-ce que "l'intepréteur publicode" si ce n'est l'instance de l'engine? 

Oui c'est bien la même chose. J'utilise "parseur" pour le code qui parse et "interpréteur" pour le code qui fait l'évaluation, car je trouve "évaluateur" pas très français dans ce contexte. 

> Je ne suis pas sûr de comprendre le "tel quel": veux-tu dire qu'il ne doit pas s'agit d'un autre engine? En gros que engine est (très grossomodo) un singleton?

Je veux dire que si on en fait un paramètre normal, il faudra toujours appeler la fonction `evaluateNode` avec la même instance de l'`engine` et pas quelque chose d'autre :

```ts
function evaluateArrondi(engine: Engine, node: ParsedNode<Arrondi>) {
  evaluateNode(engine, node.decimals) // et pas autre chose que engine
}
```

c'est pour ça que vu le design de Javascript, je ne trouve pas déconnant d'utiliser `this` dont le rôle est précisément d'être un paramètre transmis aux méthodes appelées

```ts
function evaluateArrondi(this: Engine, node: ParsedNode<Arrondi>) {
  this.evaluateNode(node.decimals) // pas moyen de changer l'instance d'engine, sauf via son interface objet
}
```

ici j'utilise un paramètre `this` explicite https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters. Dans le code actuel ce `this: Engine` est seulement visible dans la signature de la function `evaluationFunction` (et bien pris en compte par Intellisense).

Je ne pense pas que c'est la forme finale de l'API, mais l'idée était de regrouper tous les arguments dans un objet contexte, et de passer ce contexte dans l'ensemble de l'arbre enfant. Je suis ouvert à d'autres API pour rendre ces idées plus explicites.

Reply:
Ok oui c'est une bonne idée de faire usage du `this: Engine`. Je trouve ca un peu bizarre mais je pourrai m'y faire et ça suffira pour rendre les choses explicites!

Je propose un commit?

Reply:
Ok au temps pour moi.